aten/src/ATen/Allocator.h

```diff
@@ -3,14 +3,153 @@
+// is that the data pointer (e.g., float*) which you are interested in, is not
+// the same as the metadata pointer (e.g., DLManagedTensor) which you need
+// to actually deallocate the data.  Under a conventional deleter design, you
+// have to store extra context in the deleter (the metadata pointer) so that
+// you can actually delete the right thing.  Implementing this with standard
+// C++ is somewhat error-prone: if you use a std::unique_ptr to manage tensors,
+// the deleter will not be called if the data pointer is nullptr, which can
+// cause a leak if the metadata pointer is non-null (and the deleter is
+// responsible for freeing both the data pointer and the metadata pointer).
+//
+// We take a different approach.  The "metadata supervisor" situation is common
+// enough that we have organized our deleter strategy entirely around it:
+// instead of trying to make the deleter for the data pointer handle all the
+// heavy lifting, the data pointer is *non-owning*, and instead there is a
+// (type-erased) supervisor pointer which actually handles deletion.  For simple
+// cases, the supervisor pointer is the same as the data pointer, but if
+// there is some extra metadata, the supervisor pointer points there.
+//
+// There is something of a pattern to writing these; check THAllocator.{h,cpp}
+// for some examples.
+
+using DeleterFnPtr = void(*)(void*);
+using SupervisorPtr = std::unique_ptr<void, DeleterFnPtr>;
+
+// Does not delete anything
+AT_API void deleteNothing(void*);
+// Mints a SupervisorPtr that doesn't do anything.  You must
+// use this, not a nullptr, when you want a view.  However,
+// this compares equal to nullptr.
+AT_API SupervisorPtr nonOwningSupervisorPtr();
+
+class SupervisedPtr {
+private:
+  // Lifetime tied to supervisor_
+  void* data_;
+  SupervisorPtr supervisor_;
+public:
+  SupervisedPtr() : data_(nullptr), supervisor_(nonOwningSupervisorPtr()) {}
+  SupervisedPtr(void* data, SupervisorPtr&& supervisor)
+    : data_(data), supervisor_(std::move(supervisor)) {}
+  void* operator->() const { return data_; }
+  void* get() const { return data_; }
+  void* get_supervisor() const { return supervisor_.get(); }
+  void* release_supervisor() { return supervisor_.release(); }
+  template <typename T>
+  T* cast_supervisor(DeleterFnPtr expected_deleter) const {
+    if (get_deleter() != expected_deleter) return nullptr;
+    return static_cast<T*>(get_supervisor());
+  }
+  operator bool() const { return data_ || supervisor_; }
+  DeleterFnPtr get_deleter() const { return supervisor_.get_deleter(); }
+};
+
+// A DevicePtr is a SupervisedPtr that also records what device it's memory
+// is stored on.
+
+class DevicePtr {
+private:
+  // Lifetime tied to supervisor_
+  SupervisedPtr sptr_;
+  Device device_;
+public:
+  // Choice of CPU here is arbitrary; if there's an "undefined" device
+  // we could use that too
+  DevicePtr() : sptr_(), device_(kCPU) {}
```

Comment:
this seems strange, as it's going to be created with the index -1, but that's not useful later when you want to know what device this is on.

Reply:
Maybe you should validate this is a real device?  I don't remember how this works with CPU device, but there should be code that does this already.